src/modules/fancyzones/lib/ZoneWindowDrawing.cpp

```diff
@@ -11,155 +11,295 @@ namespace NonLocalizable
-        quad->rgbReserved = alpha;
-        quad->rgbRed = GetRValue(color) * alpha / 255;
-        quad->rgbGreen = GetGValue(color) * alpha / 255;
-        quad->rgbBlue = GetBValue(color) * alpha / 255;
+        return 1.f;
     }
 
-    void FillRectARGB(wil::unique_hdc& hdc, RECT const* prcFill, BYTE alpha, COLORREF color, bool blendAlpha)
+    auto tNow = std::chrono::steady_clock().now();
+    auto alpha = (tNow - *m_tAnimationStart).count() / (1e6f * m_animationDuration);
+    if (alpha < 1.f)
     {
-        BITMAPINFO bi;
-        ZeroMemory(&bi, sizeof(bi));
-        bi.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
-        bi.bmiHeader.biWidth = 1;
-        bi.bmiHeader.biHeight = 1;
-        bi.bmiHeader.biPlanes = 1;
-        bi.bmiHeader.biBitCount = 32;
-        bi.bmiHeader.biCompression = BI_RGB;
-
-        RECT fillRect;
-        CopyRect(&fillRect, prcFill);
-
-        RGBQUAD bitmapBits;
-        InitRGB(&bitmapBits, alpha, color);
-        StretchDIBits(
-            hdc.get(),
-            fillRect.left,
-            fillRect.top,
-            fillRect.right - fillRect.left,
-            fillRect.bottom - fillRect.top,
-            0,
-            0,
-            1,
-            1,
-            &bitmapBits,
-            &bi,
-            DIB_RGB_COLORS,
-            SRCCOPY);
+        return alpha;
     }
+    else
+    {
+        return 1.f;
+    }
+}
 
-    BYTE OpacitySettingToAlpha(int opacity)
+ID2D1Factory* ZoneWindowDrawing::GetD2DFactory()
+{
+    static ID2D1Factory* pD2DFactory = nullptr;
+    if (!pD2DFactory)
     {
-        return static_cast<BYTE>(opacity * 2.55);
+        D2D1CreateFactory(D2D1_FACTORY_TYPE_MULTI_THREADED, &pD2DFactory);
     }
+    return pD2DFactory;
+}
 
-    void DrawIndex(wil::unique_hdc& hdc, FancyZonesUtils::Rect rect, size_t index)
+IDWriteFactory* ZoneWindowDrawing::GetWriteFactory()
+{
+    static IUnknown* pDWriteFactory = nullptr;
+    if (!pDWriteFactory)
     {
-        Gdiplus::Graphics g(hdc.get());
+        DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &pDWriteFactory);
+    }
+    return reinterpret_cast<IDWriteFactory*>(pDWriteFactory);
+}
+
+D2D1_COLOR_F ZoneWindowDrawing::ConvertColor(COLORREF color)
+{
+    return D2D1::ColorF(GetRValue(color) / 255.f,
+                        GetGValue(color) / 255.f,
+                        GetBValue(color) / 255.f,
+                        1.f);
+}
 
-        Gdiplus::FontFamily fontFamily(NonLocalizable::SegoeUiFont);
-        Gdiplus::Font font(&fontFamily, 80, Gdiplus::FontStyleRegular, Gdiplus::UnitPixel);
-        Gdiplus::SolidBrush solidBrush(Gdiplus::Color(255, 0, 0, 0));
+D2D1_RECT_F ZoneWindowDrawing::ConvertRect(RECT rect)
+{
+    return D2D1::RectF((float)rect.left, (float)rect.top, (float)rect.right, (float)rect.bottom);
+}
+
+ZoneWindowDrawing::ZoneWindowDrawing(HWND window)
+{
+    m_window = window;
+    m_renderTarget = nullptr;
+    m_animationDuration = 0;
```

Comment:
maybe initialize it to 1 or o/w prevent division by 0 on line 23 if some future change accidently does it?

Reply:
Nice catch, I was about to do that and then I forgot to.
I solved this by explicitly coupling these two variables and wrapping them in `std::optional`